homeassistant/components/hdmi_cec.py

```diff
@@ -1,121 +1,290 @@
+        vol.Optional(CMD_DOWN): vol.Coerce(int),
+        vol.Optional(CMD_MUTE): None,
+        vol.Optional(CMD_UNMUTE): None,
+        vol.Optional(CMD_MUTE_TOGGLE): None
+    })
+}, extra=vol.REMOVE_EXTRA)
 
-def parse_mapping(mapping, parents=None):
-    """Parse configuration device mapping."""
-    if parents is None:
-        parents = []
-    for addr, val in mapping.items():
-        cur = parents + [str(addr)]
-        if isinstance(val, dict):
-            yield from parse_mapping(val, cur)
-        elif isinstance(val, str):
-            yield (val, cur)
+SERVICE_POWER_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(ATTR_DST): vol.Coerce(int),
+    })
+}, extra=vol.REMOVE_EXTRA)
 
 
-def pad_physical_address(addr):
-    """Right-pad a physical address."""
-    return addr + ['0'] * (MAX_DEPTH - len(addr))
+def setup(hass: HomeAssistant, base_config):
+    """Setup CEC capability."""
+    from pycec.network import HDMINetwork
+    from pycec import CecConfig
+    from pycec.commands import CecCommand, KeyReleaseCommand, KeyPressCommand
+    from pycec.const import KEY_VOLUME_UP, KEY_VOLUME_DOWN, KEY_MUTE, \
+        ADDR_AUDIOSYSTEM, ADDR_BROADCAST, ADDR_UNREGISTERED
 
+    hdmi_network = HDMINetwork(config=CecConfig(name="HA"), loop=hass.loop)
 
-def setup(hass, config):
-    """Setup CEC capability."""
-    global _CEC
-
-    try:
-        import cec
-    except ImportError:
-        _LOGGER.error("libcec must be installed")
-        return False
-
-    # Parse configuration into a dict of device name to physical address
-    # represented as a list of four elements.
-    flat = {}
-    for pair in parse_mapping(config[DOMAIN].get(CONF_DEVICES, {})):
-        flat[pair[0]] = pad_physical_address(pair[1])
-
-    # Configure libcec.
-    cfg = cec.libcec_configuration()
-    cfg.strDeviceName = 'HASS'
-    cfg.bActivateSource = 0
-    cfg.bMonitorOnly = 1
-    cfg.clientVersion = cec.LIBCEC_VERSION_CURRENT
-
-    # Setup CEC adapter.
-    _CEC = cec.ICECAdapter.Create(cfg)
-
-    def _power_on(call):
-        """Power on all devices."""
-        _CEC.PowerOnDevices()
-
-    def _standby(call):
-        """Standby all devices."""
-        _CEC.StandbyDevices()
-
-    def _select_device(call):
-        """Select the active device."""
-        path = flat.get(call.data[ATTR_DEVICE])
-        if not path:
-            _LOGGER.error("Device not found: %s", call.data[ATTR_DEVICE])
-        cmds = []
-        for i in range(1, MAX_DEPTH - 1):
-            addr = pad_physical_address(path[:i])
-            cmds.append('1f:82:{}{}:{}{}'.format(*addr))
-            cmds.append('1f:86:{}{}:{}{}'.format(*addr))
-        for cmd in cmds:
-            _CEC.Transmit(_CEC.CommandFromString(cmd))
-        _LOGGER.info("Selected %s", call.data[ATTR_DEVICE])
+    @callback
+    def _volume(call):
+        """Increase/decrease volume and mute/unmute system."""
+        for cmd, att in call.data.items():
+            att = int(att)
```

Comment:
Once you have the SCHEMAs validating your input, this type of coercing variables should not be needed